### ISSUE TITLE
Async Server Start

### ISSUE THREAD

**Issue by ghost** [Eastrall/Ether.Network#54]
Hi,

NetServer.Start() blocks the current thread. Is it possible to Start the server asynchronously or do I have to implement that myself by passing all that server stuff into a separate thread?

Thanks in advance.

**Comment by Eastrall** [comment]
In fact, `NetServer.Start()` blocks the current thread because of the `ManualResetEvent` (https://github.com/Eastrall/Ether.Network/blob/develop/src/Ether.Network/NetServer.cs#L94) that waits for the call of `NetServer.Stop()`.

It would be a nice feature to give the possibility to block or not the main thread when we call `NetServer.Start()`. This will allow you to implement a input with `Console.ReadLine()`.

**Comment by Eastrall** [comment]
PR #56 should solve your issue. Let me know if that's what you are looking for.

**Comment by Yothri** [comment]
Thanks, I am going to try it out soon.

**Comment by Yothri** [comment]
Hi,
so I tested this out now, and starting seems to work quite fine, but when I then enter Exit in my console command handler it breaks the loop to request  for cmd commands and finally calls the NetServer.Stop().
This however results into an ObjectDisposedException.

![devenv_2018-01-15_22-34-58](https://user-images.githubusercontent.com/5815215/34962532-5d674320-fa44-11e7-974a-d47567837e2a.png)

**Comment by Eastrall** [comment]
Thank you for your feedback. I will look into this error and make a PR to fix it.

**Comment by Yothri** [comment]
By the way: I didnt even connect a client to the running server. It happens when just starting the server in Non-Blocking mode and calling Stop on it. Just a reproducing hint :)

**Comment by Eastrall** [comment]
Can you provide one or more code samples about the different issues you are referencing in your previous message please?
This will help me to track the error and fix it. Thanks in advance! 😄

**Comment by Yothri** [comment]
[ConsoleApp1.zip](https://github.com/Eastrall/Ether.Network/files/1644411/ConsoleApp1.zip)
This is the example when Starting and Stopping the Server in nonblock mode.

Just create a sampleclient, let it connect to server, and let it disconnect from server itself after it connected. then in server try access the connection in OnDisconnected sampleserver -> Will also lead to ObjectDisposedException.

**Comment by Eastrall** [comment]
Oh I see. This is because when we disconnect the client, we first dispose it and then we call the `OnClientDisconnected` method...
We should first call the `OnClientDisconnected` and then dispose. I'll make a fix for that. Thank you for the sample!

**Comment by Eastrall** [comment]
I have fixed on my local branch the 2nd issue about the `ObjectDisposedException` on the `OnClientDisconnected`.
About your first issue, I have figured it out why it was causing the `ObjectDisposedException`. It going to be solved on next PR.

**Comment by Yothri** [comment]
Nice, would be great to have both fixed soon, so I can finally use my Console Commands hah 👍 
Thanks for your effort.

**Comment by Yothri** [comment]
I fixed that first issue by adding two lines to `StartAccept`.
It looks like this now:

```
private void StartAccept(SocketAsyncEventArgs e)
        {
            if (e.AcceptSocket != null)
                e.AcceptSocket = null;

            if (!IsRunning)
                return;

            if (!this.Socket.AcceptAsync(e))
                this.ProcessAccept(e);
        }
```

**Comment by Eastrall** [comment]
This issue should be solved with PR #62. Let me know if it's solved.
